blog/2020/05/rr.md

```diff
@@ -0,0 +1,452 @@
+   upload by clicking on a link (GitHub-based authentication is used for abuse
+   protection). She then gets back a link to include when filing a bug report
+   against julia or some other package.
+
+3. Any developer can use the link to obtain the recording and analyze it on
+   their own machine.
+
+In addition to this manual mechanism, we are also switching our linux CI systems
+to automatically create `rr` traces of any execution. That way, if a CI run fails,
+we are guaranteed to be able to debug it.
+
+If a bug report includes a link to an `rr` trace, no further reproduction
+instructions are required (if the bug is something non-obvious like
+unexpected behavior, some comments on what the expected behavior was may
+still be helpful), since the `rr` trace is guaranteed to perfectly capture
+the environment the bug was reproduced in. This almost immediately knocks
+out all the common problems I started this post with. "Works for me" is no
+longer an available answer. If it's in the trace, it broke on somebody's machine
+and can be debugged. "Heisenbug"s are no longer an issue. If it was captured
+once in `rr`, it can always be debugged. It even solves the busy-expert problem,
+as it allows non-experts to help out with triage. If such a report does not
+contain an `rr` trace, any developer, and in particular non-experts can attempt to
+reproduce the bug and create one. Even if the expert is still required to do
+the final diagnosis, doing so from an `rr` trace is orders of magnitude faster
+than from a plain bug report.
+
+# Chronomancy for dummies
+
+A computer is fundamentally a deterministic machine. Given equivalent
+states as inputs, most instructions will produce a deterministic state
+as output. So where do all the subtle execution differences come from that
+produce diverging executions and prevent bugs from reproducing? Well, the
+complete answer is complicated and has lots of details, but is roughly split
+up into:
+
+1) The input state is large. Probably the state of your entire hard drive and
+    memory. That's at least `2^2^large` bits of state right there each of which
+    could potentially cause a difference in execution. You really want to figure
+    out what state is relevant, since `2^2^large` is probably too many bits to send
+    somebody.
+
+2) Any input or other asynchronous events. This can mean user input, or
+    data from the network (via the NICs) or other devices. It also means things like
+    the timing of interrupts.
+
+3) Execution ordering and data races in multi-threaded executions.
+
+However, in theory if a tool was able to capture 100% of the relevant state
+from these categories, it could repeated generate exactly the same memory image.
+This is not a novel idea, but the devil is in the detail. As already mentioned,
+just capturing everything is probably too large, if at all possible (e.g. how
+do you capture interrupt timing - how do you even talk about time in this setting?).
```

Comment:
```suggestion
do you capture interrupt timing - how do you even talk about time in this setting?).
```

We could lie about the rate of passage of time to it?

The other possible problem here is that multithreading lies outside of the total input state space (cannot be enumerated by a Turing machine).

Reply:
The point I wanted to make was slightly different. It's that the notion of time in an execution is hard to define in such a way that makes it useful. Real time is useless, because instruction issue frequency is not anywhere near to constant. Executed instructions are fine, but may not be available in hardware etc. I should probably just say that explicitly.